File: implementation/scipamato/public/web/src/test/java/ch/difty/scipamato/publ/web/paper/browse/PublicPaperProviderTest.java
package ch.difty.scipamato.publ.web.paper.browse;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import org.apache.wicket.extensions.markup.html.repeater.data.sort.SortOrder;
import org.apache.wicket.model.IModel;
import org.apache.wicket.util.tester.WicketTester;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import ch.difty.scipamato.common.persistence.paging.PaginationContext;
import ch.difty.scipamato.common.persistence.paging.PaginationContextMatcher;
import ch.difty.scipamato.publ.ScipamatoPublicApplication;
import ch.difty.scipamato.publ.entity.PublicPaper;
import ch.difty.scipamato.publ.entity.filter.PublicPaperFilter;
import ch.difty.scipamato.publ.persistence.api.PublicPaperService;

@SuppressWarnings("SpellCheckingInspection")
@SpringBootTest
class PublicPaperProviderTest {

    private PublicPaperProvider provider;

    @Mock
    private PublicPaperService serviceMock;

    @Mock
    private PublicPaperFilter filterMock;

    @Autowired
    private ScipamatoPublicApplication application;

    private final List<PublicPaper> papers = new ArrayList<>();

    @BeforeEach
    void setUp() {
        new WicketTester(application);
        provider = new PublicPaperProvider(filterMock, 20);
        provider.setService(serviceMock);

        papers.add(new PublicPaper(1L, 1L, 1000, "authors1", "auths", "title1", "location1", "journal1", 2016, "goals1",
            "methods1", "population1", "result1", "comment1"));
        papers.add(new PublicPaper(2L, 2L, 1002, "authors2", "auths", "title2", "location2", "journal2", 2017, "goals2",
            "methods2", "population2", "result2", "comment2"));

        when(serviceMock.countByFilter(filterMock)).thenReturn(2);
        when(serviceMock.findPageByFilter(eq(filterMock), isA(PaginationContext.class))).thenReturn(papers);
    }

    @AfterEach
    void tearDown() {
        verifyNoMoreInteractions(serviceMock);
    }

    @Test
    void construct() {
        assertThat(provider.getFilterState()).isEqualTo(filterMock);
        assertThat(provider.getRowsPerPage()).isEqualTo(20);
    }

    @Test
    void construct_withNullFilter_instantiatesNewFilter() {
        PublicPaperProvider provider2 = new PublicPaperProvider(null, 10);
        assertThat(provider2.getFilterState())
            .isNotNull()
            .isNotEqualTo(filterMock)
            .isInstanceOf(PublicPaperFilter.class);
        assertThat(provider2.getRowsPerPage()).isEqualTo(10);
    }

    @Test
    void canSetFilterState() {
        assertThat(provider.getFilterState()).isEqualTo(filterMock);
        provider.setFilterState(new PublicPaperFilter());
        assertThat(provider.getFilterState())
            .isNotEqualTo(filterMock)
            .isInstanceOf(PublicPaperFilter.class);
    }

    @Test
    void newModel() {
        PublicPaper pp = PublicPaper
            .builder()
            .id(5L)
            .build();
        IModel<PublicPaper> model = provider.model(pp);
        assertThat(model).isNotNull();
        assertThat(model.getObject())
            .isNotNull()
            .isInstanceOf(PublicPaper.class);
        assertThat(model
            .getObject()
            .getId()).isEqualTo(5L);
    }

    @Test
    void gettingIterator_withAscendingSort() {
        provider.setSort("title", SortOrder.ASCENDING);
        assertThat(provider.iterator(0L, 10L))
            .extracting("id")
            .containsExactlyInAnyOrder(1L, 2L);
        verify(serviceMock).findPageByFilter(eq(filterMock), isA(PaginationContext.class));
    }

    @Test
    void gettingIterator_withDescendingSort() {
        provider.setSort("title", SortOrder.DESCENDING);
        assertThat(provider.iterator(0L, 10L))
            .extracting("id")
            .containsExactlyInAnyOrder(2L, 1L);
        verify(serviceMock).findPageByFilter(eq(filterMock), isA(PaginationContext.class));
    }

    @Test
    void gettingSize() {
        assertThat(provider.size()).isEqualTo(2);
        verify(serviceMock).countByFilter(filterMock);
    }

    @Test
    void findingAllNumbersByFilter_descendingSort() {
        findingAllNumbersByFilter(SortOrder.DESCENDING);
    }

    @Test
    void findingAllNumbersByFilter_ascendingSort() {
        findingAllNumbersByFilter(SortOrder.ASCENDING);
    }

    private void findingAllNumbersByFilter(final SortOrder srt) {
        final String sortDescription = "title: " + (SortOrder.DESCENDING.equals(srt) ? "DESC" : "ASC");
        provider.setSort("title", srt);

        when(serviceMock.findPageOfNumbersByFilter(eq(filterMock),
            argThat(new PaginationContextMatcher(0, Integer.MAX_VALUE, sortDescription)))).thenReturn(
            Arrays.asList(5L, 3L, 17L));

        assertThat(provider.findAllPaperNumbersByFilter()).containsExactly(5L, 3L, 17L);

        verify(serviceMock).findPageOfNumbersByFilter(eq(filterMock),
            argThat(new PaginationContextMatcher(0, Integer.MAX_VALUE, sortDescription)));
    }

}

File: public/public-web/src/test/kotlin/ch/difty/scipamato/publ/web/paper/browse/PublicPaperProviderTest.kt
package ch.difty.scipamato.publ.web.paper.browse

import ch.difty.scipamato.publ.ScipamatoPublicApplication
import ch.difty.scipamato.publ.entity.PublicPaper
import ch.difty.scipamato.publ.entity.filter.PublicPaperFilter
import ch.difty.scipamato.publ.persistence.api.PublicPaperService
import ch.difty.scipamato.publ.web.matchPaginationContext
import io.mockk.confirmVerified
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.verify
import org.amshove.kluent.shouldBeEqualTo
import org.amshove.kluent.shouldBeInstanceOf
import org.amshove.kluent.shouldContainSame
import org.amshove.kluent.shouldNotBeEqualTo
import org.amshove.kluent.shouldNotBeNull
import org.apache.wicket.extensions.markup.html.repeater.data.sort.SortOrder
import org.apache.wicket.util.tester.WicketTester
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest

@SpringBootTest
class PublicPaperProviderTest {

    private lateinit var provider: PublicPaperProvider

    @MockK
    private lateinit var serviceMock: PublicPaperService

    @Autowired
    private lateinit var application: ScipamatoPublicApplication

    private val filterDummy = PublicPaperFilter(number = 1L)

    private val papers: MutableList<PublicPaper> = ArrayList()

    @BeforeEach
    fun setUp() {
        WicketTester(application)
        provider = PublicPaperProvider(filterDummy, 20)
        provider.setService(serviceMock)
        papers.add(
            PublicPaper(
                1L, 1L, 1000, "authors1", "auths", "title1", "location1", "journal1", 2016, "goals1",
                "methods1", "population1", "result1", "comment1"
            )
        )
        papers.add(
            PublicPaper(
                2L, 2L, 1002, "authors2", "auths", "title2", "location2", "journal2", 2017, "goals2",
                "methods2", "population2", "result2", "comment2"
            )
        )
        every { serviceMock.countByFilter(filterDummy) } returns 2
        every { serviceMock.findPageByFilter(filterDummy, any()) } returns papers
    }

    @AfterEach
    fun tearDown() {
        confirmVerified(serviceMock)
    }

    @Test
    fun construct() {
        provider.filterState shouldBeEqualTo filterDummy
        provider.rowsPerPage shouldBeEqualTo 20
    }

    @Test
    fun construct_withNullFilter_instantiatesNewFilter() {
        val provider2 = PublicPaperProvider(null, 10)
        with(provider2.filterState) {
            shouldNotBeNull()
            this shouldBeInstanceOf PublicPaperFilter::class
            this shouldNotBeEqualTo filterDummy
        }
        provider2.rowsPerPage shouldBeEqualTo 10
    }

    @Test
    fun canSetFilterState() {
        provider.filterState shouldBeEqualTo filterDummy
        provider.filterState = PublicPaperFilter()
        provider.filterState shouldNotBeEqualTo filterDummy
        provider.filterState shouldBeInstanceOf PublicPaperFilter::class
    }

    @Test
    fun newModel() {
        val pp = PublicPaper(id = 5L)
        val model = provider.model(pp)

        model.shouldNotBeNull()
        model.`object` shouldBeInstanceOf PublicPaper::class
        model.`object`.id shouldBeEqualTo 5L
    }

    @Test
    fun gettingIterator_withAscendingSort() {
        provider.setSort("title", SortOrder.ASCENDING)
        provider.iterator(0L, 10L).asSequence().map { it.id } shouldContainSame listOf(1L, 2L).asSequence()
        verify { serviceMock.findPageByFilter(filterDummy, any()) }
    }

    @Test
    fun gettingIterator_withDescendingSort() {
        provider.setSort("title", SortOrder.DESCENDING)
        provider.iterator(0L, 10L).asSequence().map { it.id } shouldContainSame listOf(2L, 1L).asSequence()
        verify { serviceMock.findPageByFilter(filterDummy, any()) }
    }

    @Test
    fun gettingSize() {
        provider.size() shouldBeEqualTo 2
        verify { serviceMock.countByFilter(filterDummy) }
    }

    @Test
    fun findingAllNumbersByFilter_descendingSort() {
        findingAllNumbersByFilter(SortOrder.DESCENDING)
    }

    @Test
    fun findingAllNumbersByFilter_ascendingSort() {
        findingAllNumbersByFilter(SortOrder.ASCENDING)
    }

    private fun findingAllNumbersByFilter(srt: SortOrder) {
        val sortDescription = "title: " + if (SortOrder.DESCENDING == srt) "DESC" else "ASC"
        provider.setSort("title", srt)

        every {
            serviceMock.findPageOfNumbersByFilter(filterDummy, matchPaginationContext(0, Int.MAX_VALUE, sortDescription))
        } returns listOf(5L, 3L, 17L)
        provider.findAllPaperNumbersByFilter() shouldContainSame listOf(5L, 3L, 17L)
        verify { serviceMock.findPageOfNumbersByFilter(filterDummy, any()) }
    }
}
